nan.h

```diff
@@ -1591,39 +1592,51 @@ class Callback {
 
 /* abstract */ class AsyncProgressWorker : public AsyncWorker {
  public:
+
+  struct AsyncData {
+    AsyncData(const char* data_, size_t size_)
+        : data(new char[size_]), size(size_) {
+      memcpy(data, data_, size_);
+    }
+    ~AsyncData() {
+      delete data;
+    }
+    char* data;
+    size_t size;
+  };
+
+  typedef std::deque<AsyncData *> async_queue_t;
+
   explicit AsyncProgressWorker(Callback *callback_)
-      : AsyncWorker(callback_), asyncdata_(NULL), asyncsize_(0) {
-    async = new uv_async_t;
+      : AsyncWorker(callback_) {
     uv_async_init(
         uv_default_loop()
-      , async
+      , &async
       , AsyncProgress_
     );
-    async->data = this;
+    async.data = this;
 
     uv_mutex_init(&async_lock);
   }
 
   virtual ~AsyncProgressWorker() {
     uv_mutex_destroy(&async_lock);
-
-    if (asyncdata_) {
-      delete[] asyncdata_;
-    }
   }
 
   void WorkProgress() {
-    uv_mutex_lock(&async_lock);
-    char *data = asyncdata_;
-    size_t size = asyncsize_;
-    asyncdata_ = NULL;
-    uv_mutex_unlock(&async_lock);
-
     // Dont send progress events after we've already completed.
     if (callback) {
-        HandleProgressCallback(data, size);
+      uv_mutex_lock(&async_lock);
+      while (!queue.empty()) {
+        AsyncData *async_data = queue.front();
+        queue.pop_front();
+        uv_mutex_unlock(&async_lock);
+        HandleProgressCallback(async_data->data, async_data->size);
+        uv_mutex_lock(&async_lock);
+        delete async_data;
```

Comment:
I've move the delete before the call to `uv_mutex_lock()`, it doesn't need to be inside the critical section.